jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java

```diff
@@ -6614,36 +6618,175 @@ private static char convEndian(boolean isBigEndian, char value) {
 	 */
 	public native long getObjectSize(Object o);
 
+	public native boolean isClassFlattened(Class<?> clz);
+
+	private final <V> boolean compareAndSetFlattenedValue(Object obj, long offset, Class<?> clz, V v1, V v2) {
+		boolean result = false;
+		if (getValue(obj, offset, clz) == v1) {
+			putValue(obj, offset, clz, v2);
+			result = true;
+		}
+		return result;
+	}
+
+	private final <V> Object compareAndExchangeFlattenedValue(Object obj, long offset, Class<?> clz, V v1, V v2) {
+		Object result = getValue(obj, offset, clz);
+		if (result == v1) {
+			putValue(obj, offset, clz, v2);
+		}
+		return result;
+	}
+
+	/**
+	 * Atomically sets the primitive type at offset in obj if the parameter v1
+	 * matches the existing value in the object.
+	 * The get operation has memory semantics of getVolatile.
+	 * The set operation has the memory semantics of setVolatile.
+	 *
+	 * @param obj object into which to store the value
+	 * @param offset location to compare and store value in obj
+	 * @param clz the class of the parameters v1 and v2
+	 * @param v1 value that is expected to be in obj at offset
+	 * @param v2 value that will be set in obj at offset if compare is successful
+	 * @return boolean value indicating whether the field was updated
+	 */
 	public final <V> boolean compareAndSetValue(Object obj, long offset, Class<?> clz, V v1, V v2) {
-		throw OpenJDKCompileStubThrowError();
+		boolean result = false;
+		if (isClassFlattened(clz)) {
```

Comment:
This will work for arrays, but not for object fields. 

For fields, the native code to check if it is flatten is:
https://github.com/eclipse-openj9/openj9/blob/245cd10a22729c572ceb588b99f884cbddc49cad/runtime/oti/j9.h#L337-L339

You need a helper to get the field given the offset, then check if the field is flattened or not.

Reply:
I believe to do that I would have to iterate over every field in the class until I got to the correct offset, is that the correct way to do that or would it be too inefficient?

Reply:
Yes, that is the only way I am aware of.